tests/components/color_extractor/test_service.py

```diff
@@ -0,0 +1,319 @@
+    # as it wasn't reading the same RGB value set against it.
+    actual_hs = color_util.color_RGB_to_hs(*actual_rgb)
+    actual_rgb = color_util.color_hs_to_RGB(*actual_hs)
+
+    testing_hs = color_util.color_RGB_to_hs(*testing_rgb)
+    testing_rgb = color_util.color_hs_to_RGB(*testing_hs)
+
+    actual_red, actual_green, actual_blue = actual_rgb
+    testing_red, testing_green, testing_blue = testing_rgb
+
+    r_diff = abs(actual_red - testing_red)
+    g_diff = abs(actual_green - testing_green)
+    b_diff = abs(actual_blue - testing_blue)
+
+    return (
+        r_diff <= CLOSE_THRESHOLD
+        and g_diff <= CLOSE_THRESHOLD
+        and b_diff <= CLOSE_THRESHOLD
+    )
+
+
+@pytest.fixture(autouse=True)
+async def setup_light(hass):
+    """Configure our light component to work against for testing."""
+    assert await async_setup_component(
+        hass, LIGHT_DOMAIN, {LIGHT_DOMAIN: {"platform": "demo"}}
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+
+    # Validate starting values
+    assert state.state == STATE_ON
+    assert state.attributes.get(ATTR_BRIGHTNESS) == 180
+    assert state.attributes.get(ATTR_RGB_COLOR) == (255, 63, 111)
+
+    await hass.services.async_call(
+        LIGHT_DOMAIN,
+        LIGHT_SERVICE_TURN_OFF,
+        {ATTR_ENTITY_ID: LIGHT_ENTITY},
+        blocking=True,
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get(LIGHT_ENTITY)
+
+    assert state
+    assert state.state == STATE_OFF
+
+
+async def test_missing_url_and_path(hass):
+    """Test that nothing happens when url and path are missing."""
+    # Load our color_extractor component
+    await async_setup_component(
+        hass,
+        DOMAIN,
+        {},
+    )
+    await hass.async_block_till_done()
+
+    # Validate pre service call
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+    assert state.state == STATE_OFF
+
+    # Missing url and path attributes, should cause error log
+    service_data = {
+        ATTR_ENTITY_ID: LIGHT_ENTITY,
+    }
+
+    with pytest.raises(MultipleInvalid):
+        await hass.services.async_call(
+            DOMAIN, SERVICE_TURN_ON, service_data, blocking=True
+        )
+        await hass.async_block_till_done()
+
+    # check light is still off, unchanged due to bad parameters on service call
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+    assert state.state == STATE_OFF
+
+
+async def _async_load_color_extractor_url(hass, service_data):
+    # Load our color_extractor component
+    await async_setup_component(
+        hass,
+        DOMAIN,
+        {},
+    )
+    await hass.async_block_till_done()
+
+    # Validate pre service call
+    state = hass.states.get(LIGHT_ENTITY)
+    assert state
+    assert state.state == STATE_OFF
+
+    # Call the shared service, our above mock should return the base64 decoded fixture 1x1 pixel
+    assert await hass.services.async_call(
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6372789)